src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -1009,56 +1009,81 @@
-                                          :href]]
-                                :from [[{:select [[:key :name] :value :certname]
-                                         :from [(hcore/raw "jsonb_each(fs.volatile || fs.stable)")]}
-                                        :t]]}
-                               :facts_data]]}}
-      "certname" {:type :string
-                  :queryable? true
-                  :field :fs.certname}
-      "hash" {:type :string
-              :queryable? true
-              :field (hsql-hash-as-str :fs.hash)}
-      "producer_timestamp" {:type :timestamp
-                            :queryable? true
-                            :field :fs.producer_timestamp}
-      "producer" {:type :string
-                  :queryable? true
-                  :field :producers.name}
-      "environment" {:type :string
-                     :queryable? true
-                     :field :environments.environment}}
+                   :field :environments.environment}}
 
-     :selection {:from [[:factsets :fs]]
-                 :left-join [:environments
-                             [:= :fs.environment_id :environments.id]
-                             :producers
-                             [:= :producers.id :fs.producer_id]]}
+   :selection {:from [[:factsets :fs]]
+               :left-join [:environments
+                           [:= :fs.environment_id :environments.id]
+                           :producers
+                           [:= :producers.id :fs.producer_id]]}
 
-     :relationships (merge certname-relations
-                           {"environments" {:local-columns ["environment"]
-                                            :foreign-columns ["name"]}
-                            "producers" {:local-columns ["producer"]
-                                         :foreign-columns ["name"]}})
+   :relationships (merge certname-relations
+                         {"environments" {:local-columns ["environment"]
+                                          :foreign-columns ["name"]}
+                          "producers" {:local-columns ["producer"]
+                                       :foreign-columns ["name"]}})
+
+   :alias "factsets"
+   :entity :factsets
+   :source-table "factsets"
+   :subquery? false})
 
-     :alias "factsets"
-     :entity :factsets
-     :source-table "factsets"
-     :subquery? false}))
+(def factsets-query
+  "Query for the top level facts query"
+  (map->Query factsets-query-base))
+
+(def factsets-with-packages-query
+  "Query for factsets with the package_inventory reconstructed (used for sync)"
+  (let [original-joins (get-in factsets-query-base [:selection :left-join])
+        new-joins (conj original-joins
+                        :certnames
+                        [:= :certnames.certname :fs.certname]
+                        [{:select [[:certname_packages.certname_id :certname_id]
+                                   [:%array_agg.p.triple :packages]]
```

Comment:
This array aggregate function is producing an `[]` even when there are no packages submitted. The key needs to be omitted entirely if there are no entries in the package inventory tables for this node.